### ISSUE TITLE
AWS Feature: Support for application load balancer

### ISSUE THREAD

**Issue by ProTip** [kubernetes/kubernetes#30518]
[ELBv2](http://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) supports layer 7 load balancing. Surprisingly it's already GA and integrated with ECS. It would be great if kubernetes supported hooking up services to a specific ELB instance with path rules.

**Comment by mattjonesorg** [comment]
Would this be an Ingress Controller implementation for the Application Load Balancer?

**Comment by crizCraig** [comment]
Would this supercede #23483?

**Comment by ProTip** [comment]
IMHO I would say yes to both.

AWS is already referring to the ELB as "legacy"; Application Load Balancer looks like the future.

I would assume that this would be implemented as an ingress controller?

**Comment by tazjin** [comment]
I think an Ingress controller is the best way to go about this, similar to the GCP L7 load balancers. Would probably be part of k8s/contrib rather than this repo.

**Comment by johnwchadwick** [comment]
Would it not make sense to use ALBs in place of ELBs always, including in the case of Service type LoadBalancer? ELB still lacks support for WebSockets and HTTP/2.  Therefore, as long as ALBs can be used as drop-in replacements for ELBs without any major consequences, it seems very desirable to do so.

**Comment by danopia** [comment]
In the CloudFormation API, ALBs are actually called `ElasticLoadBalancingV2`. There's not much chance of ELBV1 getting new features. ALBs are a little cheaper too

**Comment by tazjin** [comment]
Changing the `LoadBalancer` type services to use ALBs could probably be switched based on an annotation. Any thoughts about that?

**Comment by johnwchadwick** [comment]
The more I think about this, the less sure I am about how it should work. ALBs are decidedly Level 7 LBs and only support HTTP (and /2 and websockets.) This is actually a good reason to not use them in place of ELB classic in the default case.

I took a look at the API. It looks like `go-aws-sdk` already supports ELBv2, which is nice. ELB classic instances and ALBs live in different namespaces.

Annotations to support ALB for Load Balancers may make sense as its still useful - but I think the idea to implement an ingress controller atop ELBv2 is probably the most sound idea, since ALBs do provide routing support. (I haven't tried using the Ingress resource, but looking at ALB rules vs Ingress rules, they definitely seem similar.)

**Comment by jokeyrhyme** [comment]
Is Kubernetes aware that certain tasks specifically offer HTTP services? Or does it only care about ports? Perhaps ELB v2 only makes sense in cases where we've specified (or it has been automatically detected)?

**Comment by hjacobs** [comment]
Having support for Application Load Balancer (ALB) could also solve #35187:

* register every pod in the ALB (own node port for each) --- this is not possible with Classic ELB
* forward the pod's node port to the pod (i.e. one iptables rule per pod)

=> We avoid one extra hop, preserve the source IP and avoid non-free cross-AZ data transfer (0.01 USD per GB)

**Comment by rajivm** [comment]
I don't think ELBv2/ALB could implement all the rules of an Ingress controller since it (still) does not support Header / Host based routing. ELBv2 only supports path based routing.

Currently, I am manually using ELBv2 in front of an Nginx Ingress controller to get the full functionality I need (HTTP 2 + Host routing to different services). It would be great to have direct support for ELBv2 as a LB option for all-http cases.

**Comment by sporkmonger** [comment]
And now ALB supports the AWS WAF, while ELB does not. Seems likely that this pattern of new feature releases targeting ALB primarily is likely to continue.

**Comment by hjacobs** [comment]
@rajivm yes, we have a similar approach using [Skipper](https://github.com/zalando/skipper) as HTTP router and another component to orchestrate ALBs (for SSL termination): https://github.com/zalando-incubator/kubernetes-on-aws/issues/169

**Comment by cemo** [comment]
This would be a great improvement because ALB supports AWS WAF as well whereas ELB does not.

**Comment by metral** [comment]
@justinsb Are there any updates to support the integration of AWS ALB's in k8s? Specifically, is the IngressController implementation bubbling up as the desired approach for this?

Thanks!

**Comment by hjacobs** [comment]
@metral what `IngressController` implementation are you referring to? We build our own small ALB manager (we use it together with [Skipper](https://github.com/zalando/skipper) which has [direct Kubernetes support](https://github.com/zalando-incubator/kubernetes-on-aws/blob/154224f8ff4be4ef39556ce9191cdd2a55944336/cluster/userdata-master.yaml#L1016)): https://github.com/zalando-incubator/kube-ingress-aws-controller 

ALBs don't yet directly support the host-based routing `Ingress` exposes, so you need to plug in something extra like nginx or Skipper.

**Comment by metral** [comment]
@hjacobs No implementation specifically. Was more wondering if ALB's were going to be conceptually tied to IngressController's in the future in some regard, given their close resemblance. Given the lack of header / host-based routing in ALB's, this seems to be a blocker for IngressController integration.

In other words, are there plans to integrate some of the custom work required to enable ALB's in k8s directly, as in your current solution, or is an implementation like yours the only route forward until ALB's and IngressControllers further align with one another?

**Comment by hrach** [comment]
Personally, I would suggest integrating ALB not as ingress controller, but only as the entry point to the cluster - for now. It supports:
- HTTP 2 (ELB not)
- WAF (ELB not)
- is less expensive than ELB ;-)

**Comment by johnwchadwick** [comment]
The problem is that it only supports HTTP, whereas LoadBalancer is by-default Level 3, and you can forward any TCP service on AWS right now.

**Comment by vendrov** [comment]
@johnwchadwick I'm not sure what is "Level 3", ELB does support TCP which is OSI 4.
Take a note that ALB does support WebSocket (which are blocked in HTTP in ELB).
While IMO ALB should be used as the default LB, It will be great If we could add annotation to the K8S-service specifying which type of LB to use.

**Comment by johnwchadwick** [comment]
Sorry, meant layer 4. I agree with allowing ALB through an annotation, but I don't think it makes sense to assume services are HTTP. That'd break most of my Kubernetes LoadBalancer services, which are largely not Websockets or any form of HTTP. If the LoadBalancer says `"protocol": "TCP"` it would be only reasonable to assume that it's a TCP proxy and not necessarily an HTTP one.

To be clear: I want ALBs in LoadBalancer too. However, I don't see how this could logically be a default setting.

**Comment by bigkraig** [comment]
Kubernetes could provision a ELB type when requesting a TCP protocol, and an ALB when using a HTTP protocol.

With regards to using ALBs as ingress controllers and the lack of hostname based routing, couldn't the ingress controller build an ALB per hostname to achieve the same thing?

**Comment by johnwchadwick** [comment]
The problem is that right now, as far as i know, `protocol` can only be `TCP` or `UDP`. If layer 7 options like `HTTP` and `HTTPS` were added, it would then make a lot more sense (and I'd be in favor of this, but it does make things somewhat less consistent admittedly.)

As far as using an ALB per-hostname, I think there's another problem: I think you can only have up to 10 routes. Correct me if I'm wrong, though.

**Comment by metral** [comment]
@bigkraig

**Comment by hjacobs** [comment]
I came to the conclusion that Ingress is the right abstraction for ALB (L7). We exclusively use Ingress now and only fallback to service type `LoadBalancer` if "raw" TCP is needed. If anybody is interested, here our setup: http://kubernetes-on-aws.readthedocs.io/en/latest/admin-guide/kubernetes-in-production.html#ingress

I believe AWS will replace the "Classic ELB" with a new load balancer product for pure-TCP eventually.

**Comment by johnwchadwick** [comment]
Interestingly, the problem has shifted. Now we can in fact use rules to do host-based routing.
https://aws.amazon.com/blogs/aws/new-host-based-routing-support-for-aws-application-load-balancers/

Coupled with the increased rule limit, it would seem that ALB is now entirely suitable for implementing Ingress controller.

**Comment by hjacobs** [comment]
@johnwchadwick yes, ALB now has host based routing. I believe we will keep our [Skipper + ALB Ingress Controller setup](https://kubernetes-on-aws.readthedocs.io/en/latest/admin-guide/kubernetes-in-production.html#ingress) as the rule limit of 75 is also arbitrary (and some user has 400 (!) Ingress rules..). Also AWS charges per rule evaluation (after first 10 free rules): https://aws.amazon.com/elasticloadbalancing/applicationloadbalancer/pricing/

**Comment by hjacobs** [comment]
It might make sense to support ALB evaluated host/path rules in https://github.com/zalando-incubator/kube-ingress-aws-controller --> https://github.com/zalando-incubator/kube-ingress-aws-controller/issues/47

**Comment by tavispaquette** [comment]
+1 for ALB support (native upgrade to websocket support)
We currently manually provision ALBs to run our Meteor apps through

**Comment by bigkraig** [comment]
We have some work happening at https://github.com/coreos/alb-ingress-controller for this

**Comment by philips** [comment]
I think this issue can be closed out. An ALB Ingress Controller is available as documented on kubernetes-users https://groups.google.com/forum/#!topic/kubernetes-users/92uFskIAnrE and it is documented in the kubernetes/ingress catalog: https://github.com/kubernetes/ingress/blob/master/docs/catalog.md

**Comment by philips** [comment]
@justinsb care to close this out?

**Comment by justinsb** [comment]
Let's wait for the support in the kubernetes/ingress repo.

**Comment by davd33** [comment]
+1

**Comment by mamoit** [comment]
Is ALB as an ingress already a thing?
This issue seems to be quite idle for its importance.

**Comment by andromedarabbit** [comment]
Give a vote for https://github.com/coreos/alb-ingress-controller . This is quite a promising project. We've adopted it and it works perfectly.

**Comment by timstoop** [comment]
Although the ingress controller is nice, I'd still like to see the possibility to replace the current Service type LoadBalancer with ALBs instead of ELBs, as we prefer using a different Ingress controller that can handle a far larger whitelist than ELBs or ALBs (which is something we need). Unless I'm unaware that ALB whitelists are unlimited?

**Comment by cdenneen** [comment]
Currently type=LoadBalancer will create classic ELB. Is there plan for that to work and create ALB instead? (elbv2)? Or is the plan to create a `Deployment` and no longer a `Service`? seems a bit confusing if that is the direction.

**Comment by eduardobaitello** [comment]
The issue is currently stalled :(
Any news about this?

**Comment by MagicJohnJang** [comment]
Ingress controller with NodePort mapping to ALB Https (443 port), solved the WebSocket issue for me.

**Comment by zparnold** [comment]
/remove-lifecycle rotten
I think this could be implemented as an additional set of annotations on the LoadBalancer for example in addition to:

service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:acm:REGION:ACCOUNT:CERT_ID"
service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
service.beta.kubernetes.io/aws-load-balancer-backend-protocol: https
We could add: service.beta.kubernetes.io/aws-load-balancer-type: elb/alb

And I would be happy to implement the functionality? Any objections? Or is this a wasted effort because of something I'm not thinking of?

**Comment by dpratt** [comment]
This would be fantastic - frankly I’d prefer to keep core ALB support to just LoadBalancer services. The new TCP ALB would be even better due to the source IP preservation.

**Comment by 97turbotalon** [comment]
I'd also be very interested in the ability to configure ALB.

**Comment by jchv** [comment]
A TCP ALB is called a Network Load Balancer or NLB and as far as I know that is in Kubernetes 1.9 today.

**Comment by cdenneen** [comment]
@jchv yeah I think that eventually the "Classic ELB" will be deprecated and maybe if the default `type=LoadBalancer` can switch from elb -> nlb under the covers then we are forward compatible. Obviously elbv2 (ALB) is different so probably wouldn't be straight forward replacement here as NLB probably could be.

**Comment by hjacobs** [comment]
@cdenneen I totally agree, the Application Load Balancer (ALB) is a L-7 router and therefore clearly is suited for Ingress instead of service `type=LoadBalancer`. Use https://github.com/zalando-incubator/kube-ingress-aws-controller or https://github.com/coreos/alb-ingress-controller and be done with it :smile:

**Comment by greg-jaunt** [comment]
NLBs do not support SSL termination and are therefore not a suitable replacement for ELBs for Kubernetes Services of type LoadBalancer.

**Comment by zparnold** [comment]
Correct, @greg-jaunt I would also say an NLB is a challenge too because you have to know in advance either the ENI ID's you're binding your TCP ALB (NLB) to, or know the IP address list in advance. Our cluster size is very dynamic and is only used to serve HTTP(s) traffic so the ALB is really the only thing I'd be working on. Let me know though if you think this'd be a wasted effort. 😄

**Comment by jchv** [comment]
If I'm not mistaking NLB is supported today with the `service.beta.kubernetes.io/aws-load-balancer-type` annotation. I'm also pretty sure it uses target groups the same way ALBs do, I'm not sure I see why it would be different.

The lack of SSL termination is a loss, but on one hand I wonder if that's even a problem. Is it mandatory for a Kubernetes LoadBalancer implementation to support SSL termination? Also, it seems possible that SSL termination at the cost of losing source IP preservation could be added into a future NLB update.

It seems really like the Ingress resource is what you want for ALB. You could implement ALB on the service level, it's no less unreasonable than implementing the HTTP/HTTPS protocol modes in ELB which Kubernetes does do as well. But, the Ingress resource will give you a lot richer functionality, and it works today. Is there a serious downside to using Ingress controllers for this?

**Comment by cdenneen** [comment]
The lack of SSL on NLB is actually a deal breaker for me. So for that reason alone I would like to see `type=LoadBalancer` for AWS actually support ALB under the coverts even if for L4 basic functionality with SSL. ALB features can be utilized further for L7 if needed with `alb-ingress-controller` but the basic `type=LoadBalancer` functionality for services would continue to work as a step towards migrating away from Classic ELB.

**Comment by jchv** [comment]
I'm unsure if I understand. Application Load Balancers do not support TCP load balancing and cannot function as L4 load Balancers at all. This is made clear in the documentation for ALBs:

>An Application Load Balancer functions at the application layer, the seventh layer of the Open Systems Interconnection (OSI) model.

There is a load balancer type that runs at L4. It's called a Network Load Balancer in AWS parlance:

>A Network Load Balancer functions at the fourth layer of the Open Systems Interconnection (OSI) model.

In the control panel, the three options  (HTTP, TCP, Classic) simply map to ALB, NLB and ELB. Unless I'm horribly mistaken, there is no ALB mode that turns it into a TCP load balancer, and NLB is the only non-classic option for doing TCP load balancing.

**Comment by zparnold** [comment]
@jchv Looks like NLB is the only accepted option at present. 
``` go
// ServiceAnnotationLoadBalancerType is the annotation used on the service
// to indicate what type of Load Balancer we want. Right now, the only accepted
// value is "nlb"
const ServiceAnnotationLoadBalancerType = "service.beta.kubernetes.io/aws-load-balancer-type"
```
Since someone else when through the trouble of figuring out wiring of AWS target groups to LB config (the new way of configuring Load Balancers in AWS,) I'll just try to hack together http/https with it. 😄 My usage of the cluster (at least for inbound traffic) is almost exclusively HTTP(s).

**Comment by hjacobs** [comment]
I think it does not make sense to keep the ALB/ELBv2 open here as the existing Ingress controllers such as https://github.com/zalando-incubator/kube-ingress-aws-controller/ solve it.

/close

**Comment by BrianChristie** [comment]
FYI for reference:
SIG-AWS has taken over the CoreOS aws-alb-ingress-controller as an official SIG-AWS Subproject as of June 1st.
- Subproject Proposal: https://docs.google.com/document/d/1h6nfuE2DWTLsnJCHdJ7jq5lYVLW_tWOlDtUVNL5OGs8/view
- Subproject Repo: https://github.com/kubernetes-sigs/aws-alb-ingress-controller

**Comment by meringu** [comment]
I don't think the ALB ingress controller addresses all use cases for ALB support in Kubernetes.
- We currently use nginx-ingress (alb-ingress can not be used, as it has a different annotation API).
- Currently use an ELB with ACM SSL termination (NLB cannot be used to terminate SSL)
- We require http2 and WebSocket support (Cannot use ELB with proxy protocol on SSL backends https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/using-elb-listenerconfig-quickref.html. As this isn't supported, backend services can not detect the client scheme when only using TCP backends)

Seemingly the only way to support ACM, nginx-ingress, WebSockets, and http2 is to use an ALB in front of the nginx-ingress Kubernetes service.

Although ALBs can do Layer 7 specific routing, they don't have to. We can configure the ALB to send all traffic to the default backend. The ALB would them act like and ELB, but with http2 and WebSocket support, leaving nginx-ingress controller to do all the awesome traffic management.

I don't think we should change the annotation default to `'alb'`, but add it as an option much like `'nlb'`.

**Comment by meringu** [comment]
AWS has just announced TLS termination with NLBs: https://aws.amazon.com/blogs/aws/new-tls-termination-for-network-load-balancers/

If this was supported in Kubernetes I would have no need for ALB support.

**Comment by GiladJudes** [comment]
+1 Would like to see ALB support added for LoadBalancer type service as well

**Comment by awsfanatic** [comment]
+1

**Comment by jamroks** [comment]
+1 for ALB support added for LoadBalancer type service, through some annotation like the one for nlb 


`service.beta.kubernetes.io/aws-load-balancer-type:  alb `

**Comment by jdomag** [comment]
+1

**Comment by smallertomato** [comment]
+1

**Comment by krystan** [comment]
> I don't think ELBv2/ALB could implement all the rules of an Ingress controller since it (still) does not support Header / Host based routing. ELBv2 only supports path based routing.
> 
> Currently, I am manually using ELBv2 in front of an Nginx Ingress controller to get the full functionality I need (HTTP 2 + Host routing to different services). It would be great to have direct support for ELBv2 as a LB option for all-http cases.

This is no longer the case: https://aws.amazon.com/blogs/aws/new-advanced-request-routing-for-aws-application-load-balancers/

**Comment by Buffer0x7cd** [comment]
I think another use case can be, where cluster are already running with some other ingress controllers like Nginx or Kong, but want to utilize WAF, which is only available on ALBs. Having the support of choosing an ALB for LoadBalancer Type ( with support of some annotations) will simplify the setup

**Comment by callum-p** [comment]
:+1: I've been wanting this for a while so I can use AWS WAF. I use the NGINX ingress controller. I don't want to swap to the ALB ingress controller that doesn't have the same feature set.

**Comment by minusv** [comment]
Any updates on this? I've been trying to use ingress controllers other than alb-ingress-controller but none of the controllers provides support for alb.
Annotation for `service.beta.kubernetes.io/aws-load-balancer-type: alb` will definitely be a huge help.
